Factory functions allow singleton size

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114971
* #114380
* #114970
* #114969
* __->__ #114897
* #115748
* #115739
* #114895

Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

soulitzer committed:
[WIP] Factory functions allow singleton size

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Factory functions allow singleton size"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -73,17 +76,27 @@ def __new__(
+        self._offsets = offsets
+        self._lengths = lengths
+
+        if is_dummy:
+            # Avoid infinite recursion
+            # Initialize the _values, _offsets, etc because they are checked
+            # when dispatching to jagged funcs in torch dispatch.
+            return
```
Are you just trying to avoid the dummy NT having a singleton associated with its size? is this just to keep the mapping between singleton <-> offsets / lengths cleaner or is there an actual problem with it having one?

it'd be nice to avoid the `is_dummy` kwarg if at all possible. trying to keep the constructor interface lean n clean

soulitzer replied:
If we create the dummy at module import time, then it shouldn't be a problem. It's just that if I create the dummy for the first time upon the registration of a singleton int, or upon the construction of the first NestedTensor, then it would lead to infinite recursion. I will see if I can get rid if `is_dummy`.

soulitzer replied:
> If we create the dummy at module import time, then it shouldn't be a problem.

I take that back. The reason its tough is because we're doing - 
create dummy -> create nested tensor -> tries to register a symint for the dummy's offsets -> needs a dummy to register symint

soulitzer replied:
> trying to keep the constructor interface lean n clean

Technically its private though, so I'm kind of okay with it lol

jbschlosser commented on aten/src/ATen/core/SingletonSymNodeImpl.h:
```diff
@@ -171,6 +189,11 @@ class C10_API SingletonSymNodeImpl : public SymNodeImpl {
  private:
   int64_t val_;
   int64_t coeff_;
+  at::Tensor values_;
```
dumb Q: why do we need to store values in addition to offsets?

soulitzer replied:
Oh these are the offsets! I've been meaning to update the naming for these fields, I know its super confusing rn lol

jbschlosser replied:
I'm good with using a more general name since we don't specifically want to tie singletons -> offsets. In fact, even for NestedTensor, a singleton might be associated with `lengths` instead. I just got confused by the name values overlapping with NT's values lol

soulitzer replied:
Renamed to `data`

jbschlosser reviewed (commented):
Looking pretty cool! got a few questions about the design

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -880,3 +880,36 @@ def embedding_default(func, *args, **kwargs):
     )
+
+@register_jagged_func(
+    torch.ops.aten._nested_zeros.default,
```
> how do we usually deal with overriding factory functions for subclasses? this might be nonsense, but would a NestedTensorMode help us?

This is an interesting approach, though the main downside I think is that the user would have to manually enable the mode. 

> maybe we can use dummy.new_empty() / dummy.new_zeros() / etc. ?

Good idea, I'm going to try this approach.

soulitzer replied:
Updated to call into new_{empty,zeros,...}

soulitzer committed:
Update on "[WIP] Factory functions allow singleton size"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Factory functions allow singleton size"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
As discussed offline, the tentative plan is to move forward with storing more metadata on the singleton. And in a later PR in the stack as the data_ is reused to also store nested_sizes for the cpp nested tensor additional metadata is necessary. A cpp mapping from offsets to singleton can be avoided if we make comparisons involving singleton's returned from cpp nested tensor's returned size error.

soulitzer commented on aten/src/ATen/core/SingletonSymNodeImpl.h:
```diff
@@ -132,6 +139,18 @@ class C10_API SingletonSymNodeImpl : public SymNodeImpl {
+    return data_;
+  }
+
+  void* singleton_dummy() override {
```
I may be able to return `TensorImpl*` instead of `void*` here since TensorImpl is actually defined in c10.

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -11,11 +11,13 @@
     if tensor_symint is None:
-        tensor_symint = torch._C._get_singleton_int(_tensor_id_counter, coeff)
+        tensor_symint = torch._C._get_singleton_int(
+            _tensor_id_counter, coeff, tensor, get_nt_dummy(), sum_offsets
```
any way we can avoid storing the dummy on the singleton since we have a single global dummy per class?

jbschlosser commented on aten/src/ATen/core/SingletonSymNodeImpl.cpp:
```diff
@@ -70,7 +70,29 @@ c10::SymNode SingletonSymNodeImpl::mul(const c10::SymNode& other) {
+
+std::optional<at::Tensor> try_call_with_dummy(std::function<at::Tensor(at::Tensor)> fn, c10::SymIntArrayRef size) {
+  at::TensorImpl* ptr = nullptr;
+  for (const auto& s : size) {
```
how will this logic be expanded to support multiple singletons in the size?

soulitzer replied:
To capture some offline discussion here:
- we would have multiple global dummy subclasses
- there would be some logic looping through the sizes and based on number/data in singleton, we use the mapping to decide which global dummy subclass to use with redispatch.
- there is also the consideration of doing more advanced multiple dispatch as we see more subclasses

jbschlosser commented on aten/src/ATen/core/SingletonSymNodeImpl.h:
```diff
@@ -34,8 +35,8 @@ class TORCH_API SingletonSymNodeImpl : public SymNodeImpl {
   // the higher-level API in python instead (TODO: actually introduce that).
-  explicit SingletonSymNodeImpl(int64_t val, int64_t coeff)
-      : val_(val), coeff_(coeff) {}
+  explicit SingletonSymNodeImpl(int64_t val, int64_t coeff, at::Tensor data, at::Tensor dummy, int64_t sum_offsets)
```
can `sum_offsets` be symbolic? should we be storing SymInt here instead?

soulitzer replied:
It won't be symbolic here actually. After metafication, the symbolic singleton e.g. SymNode class in python will hold the cpp SingletonSymNode as hint, and THAT is the one holding the symbolic sum_offsets. The cpp SingletonSymNode in the hint will still hold concrete int64_t.

jbschlosser commented on torch/fx/experimental/sym_node.py:
```diff
@@ -171,6 +178,28 @@ def is_float(self):
     def is_bool(self):
         return self.pytype is bool
 
+    def is_singleton(self):
+        return self._singleton_dummy is not None or (
+            # In case singleton SymNode is not fully initialized.
+            self.hint is not None
+            and isinstance(self.hint, SymInt)
+            and self.hint.node.is_singleton()
+        )
```
this is nice; does it mean we can get rid of the `is_singleton()` definition in `torch/fx/experimental/symbolic_shapes.py`?

soulitzer replied:
Yup, thanks for the reminder :)

soulitzer replied:
Hmm unless we want a method on SymInt, going to keep the method for now, but at least now its going to be much simpler.

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -3342,6 +3342,39 @@ def test_unbind(self, device):
+                layout=torch.jagged,
+                **kwargs
+            )
+            zeros = torch.zeros(nt.shape, **kwargs)
+            ones = torch.ones(nt.shape, **kwargs)
+            empty = torch.empty(nt.shape, **kwargs)
+            full = torch.full(nt.shape, 2, **kwargs)
```
can we also get a test where parts of `nt.shape` are taken alongside other shape parts? I think that'll be a core use case for this stuff in model code. I had to do a bunch of gymnastics in SAM to get around this not existing then

jbschlosser replied:
can we also get a test where parts of `nt.shape` are taken alongside other shape parts? I think that'll be a core use case for this stuff in model code. I had to do a bunch of gymnastics in SAM to get around this not existing then

jbschlosser reviewed (commented):
Coming along nicely, thanks for all the hard work!

At a high level, I'm curious if we can simplify / generalize what a singleton sym node contains a bit. It seems there's some jagged-specific stuff there now (dummy, sum_offsets) that won't apply for C++ NT. As discussed offline, I think it'd be cool to store conceptually (tag, data) on the singleton.

soulitzer commented on aten/src/ATen/core/SingletonSymNodeImpl.h:
```diff
@@ -134,6 +139,18 @@ class TORCH_API SingletonSymNodeImpl : public SymNodeImpl {
     return coeff_;
   }
 
+  const at::Tensor& singleton_data() {
+    return data_;
+  }
+
+  c10::TensorImpl* singleton_dummy() override {
+    return dummy_.unsafeGetTensorImpl();
+  }
+
+  int64_t singleton_sum_offsets() {
+    return sum_offsets_;
+  }
```
If its okay with you I plan to explore some ways to do this as I extend singleton to use with cpp.

soulitzer commented:
> At a high level, I'm curious if we can simplify / generalize what a singleton sym node contains a bit. It seems there's some jagged-specific stuff there now (dummy, sum_offsets) that won't apply for C++ NT. As discussed offline, I think it'd be cool to store conceptually (tag, data) on the singleton.

Having some kind of structure/indirection that hides the data that is specific to the particular symint is interesting. I prefer to explore that more as I work adding cpp NT support to singleton, but let me know if you'd like to see it more fleshed out in this PR.

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Factory functions allow singleton size"


Example codegen output

`zeros_symint`
```cpp
inline at::Tensor zeros_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    
    return autograd::make_variable(dummy.new_zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::zeros_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
}
```


`empty_symint` (special case since empty has `memory_format` which new_empty does not accept)
```cpp
inline at::Tensor empty_symint(c10::SymIntArrayRef size, at::TensorOptions options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::AutoDispatchBelowADInplaceOrView guard;
  auto ret = c10::try_call_with_dummy([=](at::Tensor dummy) {
    TORCH_CHECK(memory_format == c10::nullopt);
    return autograd::make_variable(dummy.new_empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt)), /*requires_grad=*/options.requires_grad());
  }, size);
  if (ret.has_value()) {
    return ret.value();
  }
  return autograd::make_variable(at::empty_symint(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format), /*requires_grad=*/options.requires_grad());
}
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented on test/dynamo/test_subclasses.py:
```diff
@@ -1104,6 +1105,38 @@ def backend(gm, args):
+                self.assertEqual(out, fn(x))
+
+            do_check(fn1)
+            do_check(fn2)
```
Parametrization over the test case does not work for dynamo's Test class unfortunately, so I updated to do it manually here. I didn't look too deeply, but no other dynamo test uses parametrization so I imagine it is not trivial to solve.
